libr/anal/var.c

```diff
@@ -682,7 +682,7 @@ static void extract_arg(RAnal *anal, RAnalFunction *fcn, RAnalOp *op, const char
 	r_return_if_fail (anal && fcn && op && reg);
 
 	size_t i;
-	for (i = 0; i < R_ARRAY_SIZE (op->src); i++) {
+	for (i = 0; i < (R_ARRAY_SIZE (op->src) - 1); i++) {
```

Comment:
Why skip the last element?

Reply:
Trying to fix this (`AddressSanitizer: global-buffer-overflow ../libr/anal/var.c:686 in extract_arg`):
```
=================================================================
==25321==ERROR: AddressSanitizer: global-buffer-overflow on address 0x7f28c88ceac0 at pc 0x7f28c66c55a3 bp 0x7ffddb9bf670 sp 0x7ffddb9bf660
READ of size 8 at 0x7f28c88ceac0 thread T0
    #0 0x7f28c66c55a2 in extract_arg ../libr/anal/var.c:686
    #1 0x7f28c66cc97f in r_anal_extract_vars ../libr/anal/var.c:1125
    #2 0x7f28c6627c36 in fcn_recurse ../libr/anal/fcn.c:1327
    #3 0x7f28c662878c in r_anal_fcn_bb ../libr/anal/fcn.c:1364
    #4 0x7f28c6624027 in fcn_recurse ../libr/anal/fcn.c:1091
    #5 0x7f28c662878c in r_anal_fcn_bb ../libr/anal/fcn.c:1364
    #6 0x7f28c6624027 in fcn_recurse ../libr/anal/fcn.c:1091
    #7 0x7f28c662878c in r_anal_fcn_bb ../libr/anal/fcn.c:1364
    #8 0x7f28c6624100 in fcn_recurse ../libr/anal/fcn.c:1093
    #9 0x7f28c662878c in r_anal_fcn_bb ../libr/anal/fcn.c:1364
    #10 0x7f28c6624027 in fcn_recurse ../libr/anal/fcn.c:1091
    #11 0x7f28c662878c in r_anal_fcn_bb ../libr/anal/fcn.c:1364
    #12 0x7f28c662a42e in r_anal_fcn ../libr/anal/fcn.c:1478
    #13 0x7f28cbc32fe0 in __core_anal_fcn ../libr/core/canal.c:792
    #14 0x7f28cbc45c6b in r_core_anal_fcn ../libr/core/canal.c:2017
    #15 0x7f28cbe51e8f in _anal_calls ../libr/core/cmd_anal.c:7078
    #16 0x7f28cbe52898 in cmd_anal_calls ../libr/core/cmd_anal.c:7144
    #17 0x7f28cbe7977d in cmd_anal_all ../libr/core/cmd_anal.c:9601
    #18 0x7f28cbe86c88 in cmd_anal ../libr/core/cmd_anal.c:10737
    #19 0x7f28cbffe495 in r_cmd_call ../libr/core/cmd_api.c:361
    #20 0x7f28cbfbb8c8 in r_core_cmd_subst_i ../libr/core/cmd.c:3824
    #21 0x7f28cbfac0c3 in r_core_cmd_subst ../libr/core/cmd.c:2742
    #22 0x7f28cbfef84b in run_cmd_depth ../libr/core/cmd.c:6741
    #23 0x7f28cbff09d8 in r_core_cmd ../libr/core/cmd.c:6829
    #24 0x7f28cbff0c41 in r_core_cmd_lines ../libr/core/cmd.c:6860
    #25 0x7f28d3f8d516 in run_commands ../libr/main/radare2.c:262
    #26 0x7f28d3f99b72 in r_main_radare2 ../libr/main/radare2.c:1361
    #27 0x558bcd26103b in main ../binr/radare2/radare2.c:96
    #28 0x7f28d332c0b2 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x270b2)
    #29 0x558bcd260acd in _start (/home/runner/bin/radare2+0x93acd)

0x7f28c88ceac0 is located 0 bytes to the right of global variable 'base_regs' defined in '../libr/anal/p/anal_arm_cs.c:66:17' (0x7f28c88ce9e0) of size 224
0x7f28c88ceac0 is located 32 bytes to the left of global variable 'regdelta_regs' defined in '../libr/anal/p/anal_arm_cs.c:67:17' (0x7f28c88ceae0) of size 224
SUMMARY: AddressSanitizer: global-buffer-overflow ../libr/anal/var.c:686 in extract_arg
Shadow bytes around the buggy address:
  0x0fe599111d00: f9 f9 f9 f9 01 f9 f9 f9 f9 f9 f9 f9 01 f9 f9 f9
  0x0fe599111d10: f9 f9 f9 f9 00 00 00 00 01 f9 f9 f9 f9 f9 f9 f9
  0x0fe599111d20: 00 f9 f9 f9 f9 f9 f9 f9 00 00 00 00 00 00 00 00
  0x0fe599111d30: 00 00 00 f9 f9 f9 f9 f9 00 00 00 00 00 00 00 00
  0x0fe599111d40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0fe599111d50: 00 00 00 00 00 00 00 00[f9]f9 f9 f9 00 00 00 00
  0x0fe599111d60: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fe599111d70: 00 00 00 00 00 00 00 00 f9 f9 f9 f9 00 f9 f9 f9
  0x0fe599111d80: f9 f9 f9 f9 00 f9 f9 f9 f9 f9 f9 f9 00 f9 f9 f9
  0x0fe599111d90: f9 f9 f9 f9 00 00 00 00 00 00 00 00 00 00 00 00
  0x0fe599111da0: 00 00 00 00 00 00 00 00 04 f9 f9 f9 f9 f9 f9 f9
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==25321==ABORTING

-- exit status: 1
```

Reply:
RAnalOp::src size is clearly 3 and RArraySize should also return 3.  So having loop `for (i=0; i<3; i++)` shouldn't be a problem. Error says 0 bytes to the right of global variable 'base_regs' defined in anal_arm_cs.c so it isn't complete garbage and corresponding `src[i]` point to a real RAnalValue. I would say that code doing register lookup and assigning src[i]->reg has a bug  and probably assigns pointer past the end of array not this loop.

Reply:
Ok, removed this commit then.